File: video-generator/mix-lecture.ts
#!/usr/bin/env npx ts-node
/**
 * レクチャーミックススクリプト
 * 
 * スライド単位の素材を結合して最終動画を生成
 * 
 * 使い方:
 *   npx ts-node mix-lecture.ts 3-1        # 3-1のみ
 *   npx ts-node mix-lecture.ts 3-1 3-2    # 複数指定
 *   npx ts-node mix-lecture.ts --all      # 全レクチャー
 */

import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

// パス設定
const BASE_DIR = path.resolve(__dirname, '..');
const SLIDES_DIR = path.join(BASE_DIR, 'slides');
const OUTPUT_BASE = path.join(__dirname, 'output');
const PARTS_DIR = path.join(OUTPUT_BASE, 'parts');
const VIDEOS_DIR = path.join(OUTPUT_BASE, 'videos');

// 動画設定
const VIDEO_WIDTH = 1920;
const VIDEO_HEIGHT = 1080;
const SLIDE_TRANSITION_DURATION = 0.5;  // スライド間のフェード時間

interface LectureMeta {
  lectureId: string;
  slide: string;
  totalPages: number;
  totalDuration: number;
  parts: {
    pageIndex: number;
    wavPath: string;
    duration: number;
  }[];
}

/**
 * Puppeteerでスライドの各ページをスクリーンショット
 */
async function captureSlidePages(slidePath: string, outputDir: string, pageCount: number): Promise<string[]> {
  const puppeteer = require('puppeteer');
  
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  const page = await browser.newPage();
  await page.setViewport({ width: VIDEO_WIDTH, height: VIDEO_HEIGHT });
  
  const absoluteSlidePath = path.resolve(slidePath);
  const slideUrl = `file:///${absoluteSlidePath.replace(/\\/g, '/')}`;
  
  await page.goto(slideUrl, { waitUntil: 'networkidle0' });
  
  // reveal.jsの初期化を待つ
  await page.waitForTimeout(1000);
  
  const screenshots: string[] = [];
  
  for (let i = 0; i < pageCount; i++) {
    const screenshotPath = path.join(outputDir, `page_${String(i + 1).padStart(2, '0')}.png`);
    
    // スライドを指定ページに移動
    await page.evaluate((index: number) => {
      if ((window as any).Reveal) {
        (window as any).Reveal.slide(index);
      }
    }, i);
    
    await page.waitForTimeout(300);  // アニメーション待ち
    
    await page.screenshot({ path: screenshotPath });
    screenshots.push(screenshotPath);
    
    console.log(`    スクリーンショット: ページ ${i + 1}/${pageCount}`);
  }
  
  await browser.close();
  return screenshots;
}

/**
 * 1ページ分の動画を生成（画像 + 音声）
 */
function createPageVideo(
  imagePath: string,
  wavPath: string,
  duration: number,
  outputPath: string
): boolean {
  try {
    // 画像を音声の長さ分ループして動画化
    const cmd = `ffmpeg -y -loop 1 -i "${imagePath}" -i "${wavPath}" -c:v libx264 -tune stillimage -c:a aac -b:a 192k -pix_fmt yuv420p -shortest -t ${duration + 0.5} "${outputPath}"`;
    
    execSync(cmd, { stdio: 'pipe' });
    return true;
  } catch (error: any) {
    console.error(`    動画生成エラー: ${error.message}`);
    return false;
  }
}

/**
 * 複数の動画を結合
 */
function concatenateVideos(videoPaths: string[], outputPath: string): boolean {
  try {
    // 結合リストファイルを作成
    const listPath = path.join(path.dirname(outputPath), 'concat_list.txt');
    const listContent = videoPaths.map(p => `file '${p.replace(/\\/g, '/')}'`).join('\n');
    fs.writeFileSync(listPath, listContent);
    
    // ffmpegで結合
    const cmd = `ffmpeg -y -f concat -safe 0 -i "${listPath}" -c copy "${outputPath}"`;
    execSync(cmd, { stdio: 'pipe' });
    
    // リストファイル削除
    fs.unlinkSync(listPath);
    
    return true;
  } catch (error: any) {
    console.error(`    結合エラー: ${error.message}`);
    return false;
  }
}

/**
 * 1レクチャーの最終動画を生成
 */
async function mixLecture(lectureId: string): Promise<boolean> {
  console.log(`\n[${lectureId}] ミックス開始...`);
  
  const lectureDir = path.join(PARTS_DIR, lectureId);
  const metaPath = path.join(lectureDir, '_lecture.json');
  
  if (!fs.existsSync(metaPath)) {
    console.error(`  メタデータが見つかりません: ${metaPath}`);
    console.error('  先に slide-unit-generator.ts を実行してください');
    return false;
  }
  
  const meta: LectureMeta = JSON.parse(fs.readFileSync(metaPath, 'utf-8'));
  const slidePath = path.join(SLIDES_DIR, meta.slide);
  
  console.log(`  スライド: ${meta.slide}`);
  console.log(`  ページ数: ${meta.totalPages}`);
  console.log(`  総時間: ${meta.totalDuration.toFixed(1)}秒`);
  
  // 作業ディレクトリ
  const workDir = path.join(lectureDir, '_work');
  fs.mkdirSync(workDir, { recursive: true });
  
  // Step 1: スライドのスクリーンショットを取得
  console.log('  Step 1: スクリーンショット取得...');
  const screenshots = await captureSlidePages(slidePath, workDir, meta.totalPages);
  
  // Step 2: 各ページの動画を生成
  console.log('  Step 2: ページ動画生成...');
  const pageVideos: string[] = [];
  
  for (let i = 0; i < meta.parts.length; i++) {
    const part = meta.parts[i];
    const pageVideoPath = path.join(workDir, `page_${String(i + 1).padStart(2, '0')}.mp4`);
    
    console.log(`    ページ ${i + 1}/${meta.parts.length} (${part.duration.toFixed(1)}秒)...`);
    
    const success = createPageVideo(
      screenshots[i],
      part.wavPath,
      part.duration,
      pageVideoPath
    );
    
    if (!success) {
      console.error(`    ページ ${i + 1} の動画生成に失敗`);
      return false;
    }
    
    pageVideos.push(pageVideoPath);
  }
  
  // Step 3: 動画を結合
  console.log('  Step 3: 動画結合...');
  fs.mkdirSync(VIDEOS_DIR, { recursive: true });
  const finalVideoPath = path.join(VIDEOS_DIR, `${lectureId}.mp4`);
  
  const success = concatenateVideos(pageVideos, finalVideoPath);
  
  if (success) {
    console.log(`  完了: ${finalVideoPath}`);
    
    // 作業ファイルを削除（オプション）
    // fs.rmSync(workDir, { recursive: true });
    
    return true;
  }
  
  return false;
}

/**
 * メイン処理
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('使い方:');
    console.log('  npx ts-node mix-lecture.ts <lecture-id>   # 特定のレクチャー');
    console.log('  npx ts-node mix-lecture.ts --all          # 全レクチャー');
    console.log('');
    console.log('例:');
    console.log('  npx ts-node mix-lecture.ts 3-1');
    console.log('  npx ts-node mix-lecture.ts 3-1 3-2 3-3');
    process.exit(0);
  }
  
  let lectureIds: string[];
  
  if (args.includes('--all')) {
    // 全レクチャー（partsディレクトリから取得）
    lectureIds = fs.readdirSync(PARTS_DIR)
      .filter(d => fs.statSync(path.join(PARTS_DIR, d)).isDirectory())
      .filter(d => !d.startsWith('_'))
      .sort();
  } else {
    lectureIds = args;
  }
  
  console.log('=== レクチャーミックス ===');
  console.log(`対象: ${lectureIds.join(', ')}`);
  
  let successCount = 0;
  let failCount = 0;
  
  for (const lectureId of lectureIds) {
    const success = await mixLecture(lectureId);
    if (success) {
      successCount++;
    } else {
      failCount++;
    }
  }
  
  console.log('\n=== 完了 ===');
  console.log(`成功: ${successCount}, 失敗: ${failCount}`);
}

main().catch(console.error);

File: scripts/legacy/mix_lecture.py
#!/usr/bin/env python3
"""
レクチャーミックススクリプト

parts/{lecture_id}/ にあるWAVとPNGから動画を生成して結合
"""

import argparse
import json
import os
import shutil
import subprocess
import sys
from pathlib import Path

# パス設定
BASE_DIR = Path(__file__).parent.parent
SLIDES_DIR = BASE_DIR / "slides"
PARTS_DIR = BASE_DIR / "video-generator" / "output" / "parts"
VIDEOS_DIR = BASE_DIR / "video-generator" / "output" / "videos"
GDRIVE_DIR = Path("H:/マイドライブ/d.動画作成/udemy_course/video-generator/output/videos")

VIDEO_WIDTH = 1920
VIDEO_HEIGHT = 1080


def export_slide_images(slide_md: str, output_dir: Path, lecture_id: str) -> list:
    """MarpでスライドをPNG画像に変換"""
    slide_path = SLIDES_DIR / slide_md
    
    if not slide_path.exists():
        raise Exception(f"スライドが見つかりません: {slide_path}")
    
    # Marpで画像出力
    output_pattern = output_dir / f"{lecture_id}_page.png"
    cmd = ["marp", str(slide_path), "--images", "png", "-o", str(output_pattern)]
    
    result = subprocess.run(cmd, capture_output=True, text=True, shell=True)
    if result.returncode != 0:
        raise Exception(f"marp failed: {result.stderr}")
    
    # 生成されたPNGを取得
    png_files = sorted(output_dir.glob(f"{lecture_id}_page.*.png"))
    
    # ファイル名をpage_01.pngの形式にリネーム（既存ファイルは上書き）
    renamed = []
    for i, png in enumerate(png_files):
        new_name = output_dir / f"page_{i+1:02d}.png"
        png.replace(new_name)
        renamed.append(new_name)
    
    return renamed


def create_page_video(png_path: Path, wav_path: Path, output_path: Path, silence_after: float = 0):
    """PNG + WAV から動画を生成（無音部分も音声トラックを維持）"""
    
    # 音声の長さを取得
    probe_cmd = ["ffprobe", "-v", "error", "-show_entries", "format=duration",
                 "-of", "default=noprint_wrappers=1:nokey=1", str(wav_path)]
    result = subprocess.run(probe_cmd, capture_output=True, text=True)
    duration = float(result.stdout.strip())
    
    total_duration = duration + silence_after
    
    # WAVに無音を追加してから動画生成（音声トラックを途切れさせない）
    cmd = [
        "ffmpeg", "-y",
        "-loop", "1",
        "-i", str(png_path),
        "-i", str(wav_path),
        "-filter_complex", f"[1:a]apad=whole_dur={total_duration}[a]",
        "-map", "0:v",
        "-map", "[a]",
        "-c:v", "libx264",
        "-tune", "stillimage",
        "-c:a", "aac",
        "-b:a", "192k",
        "-pix_fmt", "yuv420p",
        "-t", str(total_duration),
        str(output_path)
    ]
    
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise Exception(f"ffmpeg failed: {result.stderr}")
    
    return duration


def concatenate_videos(video_paths: list, output_path: Path):
    """動画を結合"""
    list_path = output_path.parent / "concat_list.txt"
    with open(list_path, "w", encoding="utf-8") as f:
        for vp in video_paths:
            f.write(f"file '{str(vp).replace(os.sep, '/')}'\n")
    
    cmd = [
        "ffmpeg", "-y",
        "-f", "concat",
        "-safe", "0",
        "-i", str(list_path),
        "-c", "copy",
        str(output_path)
    ]
    
    result = subprocess.run(cmd, capture_output=True, text=True)
    list_path.unlink()
    
    if result.returncode != 0:
        raise Exception(f"ffmpeg concat failed: {result.stderr}")


def concat_only(lecture_id: str):
    """_work配下のページ動画を結合のみ"""
    
    print(f"=== {lecture_id} 結合のみ ===")
    
    parts_dir = PARTS_DIR / lecture_id
    work_dir = parts_dir / "_work"
    
    if not work_dir.exists():
        print(f"エラー: _workフォルダが見つかりません: {work_dir}")
        return False
    
    # ページ動画を取得
    page_videos = sorted(work_dir.glob("page_*.mp4"))
    if not page_videos:
        print(f"エラー: ページ動画が見つかりません")
        return False
    
    print(f"ページ動画: {len(page_videos)}個")
    
    # 結合
    print("結合中...")
    VIDEOS_DIR.mkdir(parents=True, exist_ok=True)
    final_path = VIDEOS_DIR / f"{lecture_id}.mp4"
    
    concatenate_videos(page_videos, final_path)
    
    print(f"完了: {final_path}")
    return True


def mix_lecture(lecture_id: str, slide_md: str):
    """WAVとPNGから動画を生成して結合"""
    
    print(f"=== {lecture_id} ミックス ===")
    
    parts_dir = PARTS_DIR / lecture_id
    work_dir = parts_dir / "_work"
    
    # _workフォルダをクリア（既存ファイルを削除）
    if work_dir.exists():
        shutil.rmtree(work_dir)
    work_dir.mkdir(exist_ok=True)
    
    # WAVファイルを取得
    wav_files = sorted(parts_dir.glob(f"{lecture_id}_page*.wav"))
    if not wav_files:
        print(f"エラー: WAVファイルが見つかりません")
        return False
    
    print(f"WAVファイル: {len(wav_files)}個")
    
    # スライド画像を生成
    print("スライド画像生成中...")
    png_files = export_slide_images(slide_md, work_dir, lecture_id)
    print(f"PNG生成: {len(png_files)}枚")
    
    if len(wav_files) != len(png_files):
        print(f"エラー: WAV({len(wav_files)})とPNG({len(png_files)})の数が一致しません")
        return False
    
    # 各ページの動画を生成
    print("ページ動画生成中...")
    page_videos = []
    
    for i, (wav_path, png_path) in enumerate(zip(wav_files, png_files)):
        page_num = i + 1
        page_video = work_dir / f"page_{page_num:02d}.mp4"
        
        # 全ページに7秒の余韻
        silence_after = 7.0
        
        duration = create_page_video(png_path, wav_path, page_video, silence_after)
        print(f"  ページ{page_num}: {duration:.1f}秒 + {silence_after:.0f}秒")
        
        page_videos.append(page_video)
    
    # 結合
    print("結合中...")
    VIDEOS_DIR.mkdir(parents=True, exist_ok=True)
    final_path = VIDEOS_DIR / f"{lecture_id}.mp4"
    
    concatenate_videos(page_videos, final_path)
    
    # Googleドライブにコピー
    print("Googleドライブにコピー中...")
    GDRIVE_DIR.mkdir(parents=True, exist_ok=True)
    gdrive_path = GDRIVE_DIR / f"{lecture_id}.mp4"
    shutil.copy2(final_path, gdrive_path)
    
    print(f"完了: {final_path}")
    print(f"コピー先: {gdrive_path}")
    return True


def main():
    parser = argparse.ArgumentParser(description="レクチャーミックス")
    parser.add_argument("lecture_id", help="レクチャーID")
    parser.add_argument("--slide", help="スライドファイル名")
    parser.add_argument("--concat-only", action="store_true", help="_work配下のページ動画を結合のみ")
    
    args = parser.parse_args()
    
    if args.concat_only:
        success = concat_only(args.lecture_id)
    else:
        if not args.slide:
            print("エラー: --slideオプションが必要です")
            sys.exit(1)
        success = mix_lecture(args.lecture_id, args.slide)
    
    if not success:
        sys.exit(1)


if __name__ == "__main__":
    main()
